// @antv/g6/lib/plugins/minimap/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Minimap = void 0;
const g_1 = require("@antv/g");
const g_canvas_1 = require("@antv/g-canvas");
const util_1 = require("@antv/util");
const constants_1 = require("../../constants");
const id_1 = require("../../utils/id");
const padding_1 = require("../../utils/padding");
const placement_1 = require("../../utils/placement");
const point_1 = require("../../utils/point");
const base_plugin_1 = require("../base-plugin");
/**
 * <zh/> 缩略图插件
 *
 * <en/> Minimap plugin
 */
class Minimap extends base_plugin_1.BasePlugin {
    static defaultOptions = {
        size: [240, 160],
        shape: 'key',
        padding: 10,
        position: 'right-bottom',
        maskStyle: {
            border: '1px solid #ddd',
            background: 'rgba(0, 0, 0, 0.1)',
        },
        containerStyle: {
            border: '1px solid #ddd',
            background: '#fff',
        },
    };
    canvas;
    constructor(context, options) {
        super(context, Object.assign({}, Minimap.defaultOptions, options));
        this.bindEvents();
    }
    bindEvents() {
        const { graph } = this.context;
        graph.on(constants_1.GraphEvent.AFTER_DRAW, this.onDraw);
        graph.on(constants_1.GraphEvent.AFTER_RENDER, this.onRender);
        graph.on(constants_1.GraphEvent.AFTER_TRANSFORM, this.onTransform);
    }
    unbindEvents() {
        const { graph } = this.context;
        graph.off(constants_1.GraphEvent.AFTER_DRAW, this.onDraw);
        graph.off(constants_1.GraphEvent.AFTER_RENDER, this.onRender);
        graph.off(constants_1.GraphEvent.AFTER_TRANSFORM, this.onTransform);
    }
    onDraw = (event) => {
        if (event?.data?.render)
            return;
        this.onRender();
    };
    onRender = (0, util_1.throttle)(() => {
        this.renderMinimap();
        this.renderMask();
    }, 32, { leading: true });
    shapes = new Map();
    /**
     * <zh/> 创建或更新缩略图
     *
     * <en/> Create or update the minimap
     */
    renderMinimap() {
        const data = this.getElements();
        const canvas = this.initCanvas();
        this.setShapes(canvas, data);
    }
    getElements() {
        const { filter } = this.options;
        const { model } = this.context;
        const data = model.getData();
        if (!filter)
            return data;
        const { nodes, edges, combos } = data;
        return {
            nodes: nodes.filter((node) => filter((0, id_1.idOf)(node), 'node')),
            edges: edges.filter((edge) => filter((0, id_1.idOf)(edge), 'edge')),
            combos: combos.filter((combo) => filter((0, id_1.idOf)(combo), 'combo')),
        };
    }
    setShapes(canvas, data) {
        const { nodes, edges, combos } = data;
        const { shape } = this.options;
        const { element } = this.context;
        if (shape === 'key') {
            const ids = new Set();
            const iterate = (datum) => {
                const id = (0, id_1.idOf)(datum);
                ids.add(id);
                const target = element.getElement(id);
                if (!target)
                    return;
                const shape = target.getShape('key');
                const cloneShape = this.shapes.get(id) || shape.cloneNode();
                cloneShape.setPosition(shape.getPosition());
                // keep zIndex / id
                if (target.style.zIndex)
                    cloneShape.style.zIndex = target.style.zIndex;
                cloneShape.id = target.id;
                if (!this.shapes.has(id)) {
                    canvas.appendChild(cloneShape);
                    this.shapes.set(id, cloneShape);
                }
                else {
                    Object.entries(shape.attributes).forEach(([key, value]) => {
                        if (cloneShape.style[key] !== value)
                            cloneShape.style[key] = value;
                    });
                }
            };
            // 注意执行顺序 / Note the execution order
            edges.forEach(iterate);
            combos.forEach(iterate);
            nodes.forEach(iterate);
            this.shapes.forEach((shape, id) => {
                if (!ids.has(id)) {
                    canvas.removeChild(shape);
                    this.shapes.delete(id);
                }
            });
            return;
        }
        const setPosition = (id, shape) => {
            const target = element.getElement(id);
            const position = target.getPosition();
            shape.setPosition(position);
            return shape;
        };
        canvas.removeChildren();
        edges.forEach((datum) => canvas.appendChild(shape((0, id_1.idOf)(datum), 'edge')));
        combos.forEach((datum) => {
            canvas.appendChild(setPosition((0, id_1.idOf)(datum), shape((0, id_1.idOf)(datum), 'combo')));
        });
        nodes.forEach((datum) => {
            canvas.appendChild(setPosition((0, id_1.idOf)(datum), shape((0, id_1.idOf)(datum), 'node')));
        });
    }
    container;
    calculatePosition() {
        const { position, size: [w, h], } = this.options;
        const { canvas } = this.context;
        const [width, height] = canvas.getSize();
        const [x, y] = (0, placement_1.parsePlacement)(position);
        return [x * (width - w), y * (height - h)];
    }
    createContainer() {
        const { container, className, size: [width, height], containerStyle, } = this.options;
        if (container) {
            return typeof container === 'string' ? document.querySelector(container) : container;
        }
        const $container = document.createElement('div');
        $container.classList.add('g6-minimap');
        if (className)
            $container.classList.add(className);
        const [x, y] = this.calculatePosition();
        Object.assign($container.style, {
            position: 'absolute',
            left: x + 'px',
            top: y + 'px',
            width: width + 'px',
            height: height + 'px',
            ...containerStyle,
        });
        return this.context.canvas.getContainer().appendChild($container);
    }
    initCanvas() {
        const { renderer, size: [width, height], } = this.options;
        if (this.canvas) {
            this.canvas.resize(width, height);
            if (renderer)
                this.canvas.setRenderer(renderer);
        }
        else {
            const dom = document.createElement('div');
            const container = this.createContainer();
            this.container = container;
            container.appendChild(dom);
            this.canvas = new g_1.Canvas({
                width,
                height,
                container: dom,
                renderer: renderer || new g_canvas_1.Renderer(),
            });
        }
        this.setCamera();
        return this.canvas;
    }
    landmarkMap = new Map();
    createLandmark(position, focalPoint, zoom) {
        const key = `${position.join(',')}-${focalPoint.join(',')}-${zoom}`;
        if (this.landmarkMap.has(key))
            return this.landmarkMap.get(key);
        const camera = this.canvas.getCamera();
        const landmark = camera.createLandmark(key, {
            position,
            focalPoint,
            zoom,
        });
        this.landmarkMap.set(key, landmark);
        return landmark;
    }
    setCamera() {
        const { canvas } = this.context;
        const camera = this.canvas?.getCamera();
        if (!camera)
            return;
        const { size: [minimapWidth, minimapHeight], padding, } = this.options;
        const [top, right, bottom, left] = (0, padding_1.parsePadding)(padding);
        const { min: boundsMin, max: boundsMax, center } = canvas.getBounds('elements');
        const boundsWidth = boundsMax[0] - boundsMin[0];
        const boundsHeight = boundsMax[1] - boundsMin[1];
        const availableWidth = minimapWidth - left - right;
        const availableHeight = minimapHeight - top - bottom;
        const scaleX = availableWidth / boundsWidth;
        const scaleY = availableHeight / boundsHeight;
        const scale = Math.min(scaleX, scaleY);
        const landmark = this.createLandmark(center, center, scale);
        camera.gotoLandmark(landmark, 0);
    }
    mask = null;
    get maskBBox() {
        const { canvas: graphCanvas } = this.context;
        const canvasSize = graphCanvas.getSize();
        const canvasMin = graphCanvas.getCanvasByViewport([0, 0]);
        const canvasMax = graphCanvas.getCanvasByViewport(canvasSize);
        const maskMin = this.canvas.canvas2Viewport((0, point_1.toPointObject)(canvasMin));
        const maskMax = this.canvas.canvas2Viewport((0, point_1.toPointObject)(canvasMax));
        const width = maskMax.x - maskMin.x;
        const height = maskMax.y - maskMin.y;
        return [maskMin.x, maskMin.y, width, height];
    }
    /**
     * <zh/> 计算遮罩包围盒
     *
     * <en/> Calculate the bounding box of the mask
     * @returns <zh/> 遮罩包围盒 | <en/> Mask bounding box
     */
    calculateMaskBBox() {
        const { size: [minimapWidth, minimapHeight], } = this.options;
        let [x, y, width, height] = this.maskBBox;
        // clamp x, y, width, height
        if (x < 0)
            (width = upper(width + x, minimapWidth)), (x = 0);
        if (y < 0)
            (height = upper(height + y, minimapHeight)), (y = 0);
        if (x + width > minimapWidth)
            width = lower(minimapWidth - x, 0);
        if (y + height > minimapHeight)
            height = lower(minimapHeight - y, 0);
        return [upper(x, minimapWidth), upper(y, minimapHeight), lower(width, 0), lower(height, 0)];
    }
    /**
     * <zh/> 创建或更新遮罩
     *
     * <en/> Create or update the mask
     */
    renderMask() {
        const { maskStyle } = this.options;
        if (!this.mask) {
            this.mask = document.createElement('div');
            this.mask.addEventListener('pointerdown', this.onMaskDragStart);
        }
        this.container.appendChild(this.mask);
        Object.assign(this.mask.style, {
            ...maskStyle,
            cursor: 'move',
            position: 'absolute',
            pointerEvents: 'auto',
        });
        this.updateMask();
    }
    isMaskDragging = false;
    onMaskDragStart = (event) => {
        if (!this.mask)
            return;
        this.isMaskDragging = true;
        this.mask.setPointerCapture(event.pointerId);
        this.mask.addEventListener('pointermove', this.onMaskDrag);
        this.mask.addEventListener('pointerup', this.onMaskDragEnd);
        this.mask.addEventListener('pointercancel', this.onMaskDragEnd);
    };
    onMaskDrag = (event) => {
        if (!this.mask || !this.isMaskDragging)
            return;
        const { size: [minimapWidth, minimapHeight], } = this.options;
        const { movementX, movementY } = event;
        const { left, top, width: w, height: h } = this.mask.style;
        const [, , fullWidth, fullHeight] = this.maskBBox;
        let x = parseInt(left) + movementX;
        let y = parseInt(top) + movementY;
        let width = parseInt(w);
        let height = parseInt(h);
        // 确保 mask 在 minimap 内部
        // Ensure that the mask is inside the minimap
        if (x < 0)
            x = 0;
        if (y < 0)
            y = 0;
        if (x + width > minimapWidth)
            x = lower(minimapWidth - width, 0);
        if (y + height > minimapHeight)
            y = lower(minimapHeight - height, 0);
        // 当拖拽画布导致 mask 缩小时，拖拽 mask 时，能够恢复到实际大小
        // When dragging the canvas causes the mask to shrink, dragging the mask will restore it to its actual size
        if (width < fullWidth) {
            if (movementX > 0)
                (x = lower(x - movementX, 0)), (width = upper(width + movementX, minimapWidth));
            else if (movementX < 0)
                width = upper(width - movementX, minimapWidth);
        }
        if (height < fullHeight) {
            if (movementY > 0)
                (y = lower(y - movementY, 0)), (height = upper(height + movementY, minimapHeight));
            else if (movementY < 0)
                height = upper(height - movementY, minimapHeight);
        }
        Object.assign(this.mask.style, {
            left: x + 'px',
            top: y + 'px',
            width: width + 'px',
            height: height + 'px',
        });
        // 基于 movement 进行相对移动
        // Move relative to movement
        const deltaX = parseInt(left) - x;
        const deltaY = parseInt(top) - y;
        if (deltaX === 0 && deltaY === 0)
            return;
        const zoom1 = this.context.canvas.getCamera().getZoom();
        const zoom2 = this.canvas.getCamera().getZoom();
        const ratio = zoom1 / zoom2;
        this.context.graph.translateBy([deltaX * ratio, deltaY * ratio], false);
    };
    onMaskDragEnd = (event) => {
        if (!this.mask)
            return;
        this.isMaskDragging = false;
        this.mask.releasePointerCapture(event.pointerId);
        this.mask.removeEventListener('pointermove', this.onMaskDrag);
        this.mask.removeEventListener('pointerup', this.onMaskDragEnd);
        this.mask.removeEventListener('pointercancel', this.onMaskDragEnd);
    };
    onTransform = (0, util_1.throttle)(() => {
        if (this.isMaskDragging)
            return;
        this.updateMask();
        this.setCamera();
    }, 32, { leading: true });
    updateMask() {
        if (!this.mask)
            return;
        const [x, y, width, height] = this.calculateMaskBBox();
        Object.assign(this.mask.style, {
            top: y + 'px',
            left: x + 'px',
            width: width + 'px',
            height: height + 'px',
        });
    }
    destroy() {
        this.unbindEvents();
        this.canvas.destroy();
        this.mask?.remove();
        super.destroy();
    }
}
exports.Minimap = Minimap;
const upper = (value, max) => Math.min(value, max);
const lower = (value, min) => Math.max(value, min);
