// @antv/g6/lib/transforms/position-radial-labels.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PositionRadialLabels = void 0;
const g_1 = require("@antv/g");
const id_1 = require("../utils/id");
const position_1 = require("../utils/position");
const size_1 = require("../utils/size");
const vector_1 = require("../utils/vector");
const base_transform_1 = require("./base-transform");
/**
 * <zh/> 根据径向布局自动调整节点标签样式，包括位置和旋转角度
 *
 * <en/> Automatically adjust the style of node labels according to the radial layout, including position and rotation angle
 */
class PositionRadialLabels extends base_transform_1.BaseTransform {
    static defaultOptions = {
        offset: 5,
    };
    constructor(context, options) {
        super(context, Object.assign({}, PositionRadialLabels.defaultOptions, options));
    }
    get ref() {
        return this.context.model.getRootsData()[0];
    }
    afterLayout() {
        const refPoint = (0, position_1.positionOf)(this.ref);
        const { graph, model } = this.context;
        const data = model.getData();
        data.nodes?.forEach((datum) => {
            if ((0, id_1.idOf)(datum) === (0, id_1.idOf)(this.ref))
                return;
            const radian = (0, vector_1.rad)((0, vector_1.subtract)((0, position_1.positionOf)(datum), refPoint));
            const isLeft = Math.abs(radian) > Math.PI / 2;
            const isLeaf = !datum.children || datum.children.length === 0;
            const nodeHalfWidth = (0, size_1.parseSize)(graph.getElementRenderStyle((0, id_1.idOf)(datum)).size)[0] / 2;
            const offset = (isLeaf ? 1 : -1) * (nodeHalfWidth + this.options.offset);
            const translate = `translate(${offset * Math.cos(radian)},${offset * Math.sin(radian)})`;
            const rotate = `rotate(${isLeft ? (0, g_1.rad2deg)(radian) + 180 : (0, g_1.rad2deg)(radian)}deg)`;
            model.updateNodeData([
                {
                    id: (0, id_1.idOf)(datum),
                    style: {
                        labelTextAlign: isLeft === isLeaf ? 'right' : 'left',
                        labelTextBaseline: 'middle',
                        labelTransform: `${translate} ${rotate}`,
                    },
                },
            ]);
        });
        graph.draw();
    }
}
exports.PositionRadialLabels = PositionRadialLabels;
