// @antv/g6/esm/elements/nodes/base-node.js
import { Circle as GCircle } from '@antv/g';
import { getPortXYByPlacement, getTextStyleByPlacement, isSimplePort } from '../../utils/element';
import { inferIconStyle } from '../../utils/node';
import { getPaletteColors } from '../../utils/palette';
import { getRectIntersectPoint } from '../../utils/point';
import { omitStyleProps, subObject, subStyleProps } from '../../utils/prefix';
import { parseSize } from '../../utils/size';
import { mergeOptions } from '../../utils/style';
import { getWordWrapWidthByBox } from '../../utils/text';
import { setVisibility } from '../../utils/visibility';
import { BaseElement } from '../base-element';
import { effect } from '../effect';
import { Badge, Icon, Label } from '../shapes';
import { connectImage, dispatchPositionChange } from '../shapes/image';
/**
 * <zh/> 节点元素的基类
 *
 * <en/> Base node class
 * @remarks
 * <zh/> 自定义节点时，建议将此类作为基类。这样，你只需要关注如何实现 keyShape 的绘制逻辑
 *
 * <zh/> 设计文档：https://www.yuque.com/antv/g6/gl1iof1xpzg6ed98
 *
 * <en/> When customizing a node, it is recommended to use this class as the base class. This way, you can directly focus on how to implement the drawing logic of keyShape
 *
 * <en/> Design document: https://www.yuque.com/antv/g6/gl1iof1xpzg6ed98
 */
export class BaseNode extends BaseElement {
    type = 'node';
    static defaultStyleProps = {
        x: 0,
        y: 0,
        size: 32,
        droppable: true,
        draggable: true,
        port: true,
        ports: [],
        portZIndex: 2,
        portLinkToCenter: false,
        badge: true,
        badges: [],
        badgeZIndex: 3,
        halo: false,
        haloDroppable: false,
        haloLineDash: 0,
        haloLineWidth: 12,
        haloStrokeOpacity: 0.25,
        haloPointerEvents: 'none',
        haloZIndex: -1,
        icon: true,
        iconZIndex: 1,
        label: true,
        labelIsBillboard: true,
        labelMaxWidth: '200%',
        labelPlacement: 'bottom',
        labelWordWrap: false,
        labelZIndex: 0,
    };
    constructor(options) {
        super(mergeOptions({ style: BaseNode.defaultStyleProps }, options));
    }
    getSize(attributes = this.attributes) {
        const { size } = attributes;
        return parseSize(size);
    }
    getKeyStyle(attributes) {
        const style = this.getGraphicStyle(attributes);
        return Object.assign(omitStyleProps(style, ['label', 'halo', 'icon', 'badge', 'port']));
    }
    getLabelStyle(attributes) {
        if (attributes.label === false || !attributes.labelText)
            return false;
        const { placement, maxWidth, offsetX, offsetY, ...labelStyle } = subStyleProps(this.getGraphicStyle(attributes), 'label');
        const keyBounds = this.getShape('key').getLocalBounds();
        return Object.assign(getTextStyleByPlacement(keyBounds, placement, offsetX, offsetY), { wordWrapWidth: getWordWrapWidthByBox(keyBounds, maxWidth) }, labelStyle);
    }
    getHaloStyle(attributes) {
        if (attributes.halo === false)
            return false;
        const { fill, ...keyStyle } = this.getKeyStyle(attributes);
        const haloStyle = subStyleProps(this.getGraphicStyle(attributes), 'halo');
        return { ...keyStyle, stroke: fill, ...haloStyle };
    }
    getIconStyle(attributes) {
        if (attributes.icon === false || (!attributes.iconText && !attributes.iconSrc))
            return false;
        const iconStyle = subStyleProps(this.getGraphicStyle(attributes), 'icon');
        return Object.assign(inferIconStyle(attributes.size, iconStyle), iconStyle);
    }
    getBadgesStyle(attributes) {
        const badges = subObject(this.shapeMap, 'badge-');
        const badgesShapeStyle = {};
        Object.keys(badges).forEach((key) => {
            badgesShapeStyle[key] = false;
        });
        if (attributes.badge === false || !attributes.badges?.length)
            return badgesShapeStyle;
        const { badges: badgeOptions = [], badgePalette, opacity = 1, ...restAttributes } = attributes;
        const colors = getPaletteColors(badgePalette);
        const badgeStyle = subStyleProps(this.getGraphicStyle(restAttributes), 'badge');
        badgeOptions.forEach((option, i) => {
            badgesShapeStyle[i] = {
                backgroundFill: colors ? colors[i % colors?.length] : undefined,
                opacity,
                ...badgeStyle,
                ...this.getBadgeStyle(option),
            };
        });
        return badgesShapeStyle;
    }
    getBadgeStyle(style) {
        const keyShape = this.getShape('key');
        const { placement = 'top', offsetX, offsetY, ...restStyle } = style;
        const textStyle = getTextStyleByPlacement(keyShape.getLocalBounds(), placement, offsetX, offsetY, true);
        return { ...textStyle, ...restStyle };
    }
    getPortsStyle(attributes) {
        const ports = this.getPorts();
        const portsShapeStyle = {};
        Object.keys(ports).forEach((key) => {
            portsShapeStyle[key] = false;
        });
        if (attributes.port === false || !attributes.ports?.length)
            return portsShapeStyle;
        const portStyle = subStyleProps(this.getGraphicStyle(attributes), 'port');
        const { ports: portOptions = [] } = attributes;
        portOptions.forEach((option, index) => {
            const key = option.key || index;
            const mergedStyle = { ...portStyle, ...option };
            if (isSimplePort(mergedStyle)) {
                portsShapeStyle[key] = false;
            }
            else {
                const [x, y] = this.getPortXY(attributes, option);
                portsShapeStyle[key] = { transform: `translate(${x}, ${y})`, ...mergedStyle };
            }
        });
        return portsShapeStyle;
    }
    getPortXY(attributes, style) {
        const { placement = 'left' } = style;
        const keyShape = this.getShape('key');
        return getPortXYByPlacement(getBoundsInOffscreen(this.context, keyShape), placement);
    }
    /**
     * Get the ports for the node.
     * @returns Ports shape map.
     */
    getPorts() {
        return subObject(this.shapeMap, 'port-');
    }
    /**
     * Get the center point of the node.
     * @returns The center point of the node.
     */
    getCenter() {
        return this.getShape('key').getBounds().center;
    }
    /**
     * Get the point on the outer contour of the node that is the intersection with a line starting in the center the ending in the point `p`.
     * @param point - The point to intersect with the node.
     * @param useExtendedLine - Whether to use the extended line.
     * @returns The intersection point.
     */
    getIntersectPoint(point, useExtendedLine = false) {
        const keyShapeBounds = this.getShape('key').getBounds();
        return getRectIntersectPoint(point, keyShapeBounds, useExtendedLine);
    }
    drawHaloShape(attributes, container) {
        const style = this.getHaloStyle(attributes);
        if (!effect(this, 'halo', style))
            return;
        const keyShape = this.getShape('key');
        this.upsert('halo', keyShape.constructor, style, container);
    }
    drawIconShape(attributes, container) {
        const style = this.getIconStyle(attributes);
        if (!effect(this, 'icon', style))
            return;
        this.upsert('icon', Icon, style, container);
        connectImage(this);
    }
    drawBadgeShapes(attributes, container) {
        const badgesStyle = this.getBadgesStyle(attributes);
        Object.keys(badgesStyle).forEach((key) => {
            const style = badgesStyle[key];
            if (!effect(this, `badge-${key}`, style))
                return;
            this.upsert(`badge-${key}`, Badge, style, container);
        });
    }
    drawPortShapes(attributes, container) {
        const portsStyle = this.getPortsStyle(attributes);
        Object.keys(portsStyle).forEach((key) => {
            const style = portsStyle[key];
            const shapeKey = `port-${key}`;
            if (!effect(this, shapeKey, style))
                return;
            this.upsert(shapeKey, GCircle, style, container);
        });
    }
    drawLabelShape(attributes, container) {
        const style = this.getLabelStyle(attributes);
        if (!effect(this, 'label', style))
            return;
        this.upsert('label', Label, style, container);
    }
    // 用于装饰抽象方法 / Used to decorate abstract methods
    _drawKeyShape(attributes, container) {
        return this.drawKeyShape(attributes, container);
    }
    render(attributes = this.parsedAttributes, container = this) {
        // 1. key shape
        this._drawKeyShape(attributes, container);
        if (!this.getShape('key'))
            return;
        // 2. halo, use shape same with keyShape
        this.drawHaloShape(attributes, container);
        // 3. icon
        this.drawIconShape(attributes, container);
        // 4. badges
        this.drawBadgeShapes(attributes, container);
        // 5. label
        this.drawLabelShape(attributes, container);
        // 6. ports
        this.drawPortShapes(attributes, container);
    }
    update(attr) {
        super.update(attr);
        if (attr && ('x' in attr || 'y' in attr || 'z' in attr)) {
            dispatchPositionChange(this);
        }
    }
    onframe() {
        this.drawBadgeShapes(this.parsedAttributes, this);
        this.drawLabelShape(this.parsedAttributes, this);
    }
}
/**
 * <zh/> 在离屏画布中获取图形包围盒
 *
 * <en/> Get the bounding box of the graphic in the off-screen canvas
 * @param context - <zh/> 运行时上下文 <en/> Runtime context
 * @param shape - <zh/> 图形实例 <en/> Graphic instance
 * @returns <zh/> 图形包围盒 <en/> Graphic bounding box
 */
function getBoundsInOffscreen(context, shape) {
    if (!context)
        return shape.getLocalBounds();
    // 将主图形靠背至全局空间，避免受到父级 transform 的影响
    // 合理的操作应该是靠背至离屏画布，但目前 G 有点问题
    // Move the main graphic to the global space to avoid being affected by the parent transform
    // The reasonable operation should be moved to the off-screen canvas, but there is a problem with G at present
    const canvas = context.canvas.getLayer();
    const substitute = shape.cloneNode();
    setVisibility(substitute, 'hidden');
    canvas.appendChild(substitute);
    const bounds = substitute.getLocalBounds();
    canvas.removeChild(substitute);
    return bounds;
}
