// @antv/g6/lib/behaviors/optimize-viewport-transform.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OptimizeViewportTransform = void 0;
const util_1 = require("@antv/util");
const constants_1 = require("../constants");
const visibility_1 = require("../utils/visibility");
const base_behavior_1 = require("./base-behavior");
/**
 * <zh/> 操作画布过程中隐藏元素
 *
 * <en/> Hide elements during canvas operations (dragging, zooming, scrolling)
 */
class OptimizeViewportTransform extends base_behavior_1.BaseBehavior {
    static defaultOptions = {
        enable: true,
        debounce: 200,
        shapes: { node: ['key'] },
    };
    isVisible = true;
    constructor(context, options) {
        super(context, Object.assign({}, OptimizeViewportTransform.defaultOptions, options));
        this.bindEvents();
    }
    setElementsVisibility = (elements, visibility, excludedClassnames) => {
        elements.forEach((element) => {
            (0, visibility_1.setVisibility)(element, visibility, false, (shape) => {
                if (!shape.className)
                    return true;
                return !excludedClassnames?.includes(shape.className);
            });
        });
    };
    hideShapes = (event) => {
        if (!this.validate(event) || !this.isVisible)
            return;
        const { element } = this.context;
        const { shapes = {} } = this.options;
        this.setElementsVisibility(element.getNodes(), 'hidden', shapes.node);
        this.setElementsVisibility(element.getEdges(), 'hidden', shapes.edge);
        this.setElementsVisibility(element.getCombos(), 'hidden', shapes.combo);
        this.isVisible = false;
    };
    showShapes = (0, util_1.debounce)((event) => {
        if (!this.validate(event) || this.isVisible)
            return;
        const { element } = this.context;
        this.setElementsVisibility(element.getNodes(), 'visible');
        this.setElementsVisibility(element.getEdges(), 'visible');
        this.setElementsVisibility(element.getCombos(), 'visible');
        this.isVisible = true;
    }, this.options.debounce);
    bindEvents() {
        const { graph } = this.context;
        graph.on(constants_1.GraphEvent.BEFORE_TRANSFORM, this.hideShapes);
        graph.on(constants_1.GraphEvent.AFTER_TRANSFORM, this.showShapes);
    }
    unbindEvents() {
        const { graph } = this.context;
        graph.off(constants_1.GraphEvent.BEFORE_TRANSFORM, this.hideShapes);
        graph.off(constants_1.GraphEvent.AFTER_TRANSFORM, this.showShapes);
    }
    validate(event) {
        if (this.destroyed)
            return false;
        const { enable } = this.options;
        if ((0, util_1.isFunction)(enable))
            return enable(event);
        return !!enable;
    }
    destroy() {
        this.unbindEvents();
        super.destroy();
    }
}
exports.OptimizeViewportTransform = OptimizeViewportTransform;
